File: docs/sphinx/snippets/cpp/using/backends/dynamics.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// Compile and run with:
// ```
// $nvq++ --target dynamics dynamics.cpp -o dynamics && ./dynamics
// ```

#include "cudaq/algorithms/evolve.h"
#include "cudaq/algorithms/integrator.h"
#include "cudaq/operators.h"
#include <cmath>
#include <complex>
#include <cudaq.h>
#include <iostream>
#include <vector>

using namespace cudaq;

int main() {
  {
    // [Begin Transmon]
    // Parameters
    double omega_z = 6.5;
    double omega_x = 4.0;
    double omega_d = 0.5;

    // Qubit Hamiltonian
    auto hamiltonian = spin_op(0.5 * omega_z * spin_op::z(0));

    // Time dependent modulation
    auto mod_func =
        [omega_d](const parameter_map &params) -> std::complex<double> {
      auto it = params.find("t");
      if (it != params.end()) {
        double t = it->second.real();
        const auto result = std::cos(omega_d * t);
        return result;
      }
      throw std::runtime_error("Cannot find the time parameter.");
    };

    hamiltonian += mod_func * spin_op::x(0) * omega_x;
    // [End Transmon]
    // [Begin Evolve]
    double t_final = 1.0;
    int n_steps = 100;

    // Define dimensions of subsystem (single two-level system)
    cudaq::dimension_map dimensions = {{0, 2}};

    // Initial state (ground state)
    auto psi0 =
        cudaq::state::from_data(std::vector<std::complex<double>>{1.0, 0.0});

    // Schedule of time steps
    std::vector<double> steps(n_steps);
    for (int i = 0; i < n_steps; i++)
      steps[i] = i * t_final / (n_steps - 1);

    schedule schedule(steps, {"t"});

    // Numerical integrator
    // Here we choose a Runge-`Kutta` method for time evolution.
    cudaq::integrators::runge_kutta integrator(4, 0.01);

    // Observables to track
    auto observables = {cudaq::spin_op::x(0), cudaq::spin_op::y(0),
                        cudaq::spin_op::z(0)};

    // Run simulation
    // We evolve the system under the defined Hamiltonian. No collapsed
    // operators are provided (closed system evolution). The evolution returns
    // expectation values for all defined observables at each time step.
    auto evolution_result =
        cudaq::evolve(hamiltonian, dimensions, schedule, psi0, integrator, {},
                      observables, cudaq::IntermediateResultSave::All);
    // [End Evolve]
    // [Begin Print]
    // Extract and print results
    for (size_t i = 0; i < steps.size(); i++) {
      double ex =
          evolution_result.expectation_values.value()[0][i].expectation();
      double ey =
          evolution_result.expectation_values.value()[1][i].expectation();
      double ez =
          evolution_result.expectation_values.value()[2][i].expectation();
      std::cout << steps[i] << " " << ex << " " << ey << " " << ez << "\n";
    }
    // [End Print]
  }

  {
    // Jaynes-Cummings Hamiltonian
    double omega_c = 6.0 * M_PI;
    double omega_a = 4.0 * M_PI;
    double Omega = 0.5;

    // [Begin Jaynes-Cummings]
    // Jaynes-Cummings Hamiltonian
    auto jc_hamiltonian =
        omega_c * boson_op::create(1) * boson_op::annihilate(1) +
        (omega_a / 2.0) * spin_op::z(0) +
        (Omega / 2.0) * (boson_op::annihilate(1) * spin_op::plus(0) +
                         boson_op::create(1) * spin_op::minus(0));
    // [End Jaynes-Cummings]

    // [Begin Hamiltonian]
    // Hamiltonian with driving frequency
    double omega = M_PI;
    auto H0 = spin_op::z(0);
    auto H1 = spin_op::x(0);
    auto mod_func =
        [omega](const std::unordered_map<std::string, std::complex<double>>
                    &parameters) {
          auto entry = parameters.find("t");
          if (entry == parameters.end())
            throw std::runtime_error("Cannot find value of expected parameter");
          const auto t = entry->second.real();
          return std::cos(omega * t);
        };
    auto driven_hamiltonian = H0 + mod_func * H1;
    // [End Hamiltonian]
  }
  {
    // [Begin DefineOp]
    auto displacement_matrix =
        [](const std::vector<int64_t> &dimensions,
           const std::unordered_map<std::string, std::complex<double>>
               &parameters) -> cudaq::complex_matrix {
      // Returns the displacement operator matrix.
      //  Args:
      //   - displacement: Amplitude of the displacement operator.
      // See also https://en.wikipedia.org/wiki/Displacement_operator.
      std::size_t dimension = dimensions[0];
      auto entry = parameters.find("displacement");
      if (entry == parameters.end())
        throw std::runtime_error("missing value for parameter 'displacement'");
      auto displacement_amplitude = entry->second;
      auto create = cudaq::complex_matrix(dimension, dimension);
      auto annihilate = cudaq::complex_matrix(dimension, dimension);
      for (std::size_t i = 0; i + 1 < dimension; i++) {
        create[{i + 1, i}] = std::sqrt(static_cast<double>(i + 1));
        annihilate[{i, i + 1}] = std::sqrt(static_cast<double>(i + 1));
      }
      auto term1 = displacement_amplitude * create;
      auto term2 = std::conj(displacement_amplitude) * annihilate;
      return (term1 - term2).exponential();
    };

    cudaq::matrix_handler::define("displace_op", {-1}, displacement_matrix);

    // Instantiate a displacement operator acting on the given degree of
    // freedom.
    auto displacement = [](std::size_t degree) {
      return cudaq::matrix_handler::instantiate("displace_op", {degree});
    };
    // [End DefineOp]

    // [Begin Schedule1]

    // Define a system consisting of a single degree of freedom (0) with
    // dimension 3.
    cudaq::dimension_map system_dimensions{{0, 3}};
    auto system_operator = displacement(0);

    // Define the time dependency of the system operator as a schedule that
    // linearly increases the displacement parameter from 0 to 1.
    cudaq::schedule time_dependence(cudaq::linspace(0, 1, 100),
                                    {"displacement"});
    const std::vector<std::complex<double>> state_vec(3, 1.0 / std::sqrt(3.0));
    auto initial_state = cudaq::state::from_data(state_vec);
    cudaq::integrators::runge_kutta integrator(4, 0.01);
    // Simulate the evolution of the system under this time dependent operator.
    cudaq::evolve(system_operator, system_dimensions, time_dependence,
                  initial_state, integrator);
    // [End Schedule1]

    // [Begin Schedule2]
    auto hamiltonian = displacement(0) + cudaq::matrix_op::squeeze(0);

    // Define a schedule such that displacement amplitude increases linearly in
    // time but the squeezing amplitude decreases, that is follows the inverse
    // schedule. def parameter_values(time_steps):
    auto parameter_values = [](const std::vector<double> &time_steps) {
      auto compute_value = [time_steps](const std::string &param_name,
                                        const std::complex<double> &step) {
        int step_idx = (int)step.real();
        if (param_name == "displacement")
          return time_steps[step_idx];
        if (param_name == "squeezing")
          return time_steps[time_steps.size() - (step_idx + 1)];

        throw std::runtime_error("value for parameter " + param_name +
                                 " undefined");
      };

      std::vector<std::complex<double>> steps;
      for (int i = 0; i < time_steps.size(); ++i)
        steps.emplace_back(i);
      return cudaq::schedule(steps, {"displacement", "squeezing"},
                             compute_value);
    };

    auto time_dependence_param = parameter_values(cudaq::linspace(0, 1, 100));
    cudaq::evolve(hamiltonian, system_dimensions, time_dependence_param,
                  initial_state, integrator);
    // [End Schedule2]
  }
  {
    // [Begin SuperOperator]
    const cudaq::dimension_map dims = {{0, 2}};
    cudaq::product_op<cudaq::matrix_handler> ham_ =
        2.0 * M_PI * 0.1 * cudaq::spin_op::x(0);
    cudaq::sum_op<cudaq::matrix_handler> ham(ham_);
    constexpr int numSteps = 10;
    cudaq::schedule schedule(cudaq::linspace(0.0, 1.0, numSteps), {"t"});
    auto initialState =
        cudaq::state::from_data(std::vector<std::complex<double>>{1.0, 0.0});
    cudaq::integrators::runge_kutta integrator(1, 0.001);
    // Create a super-operator to evolve the system under the Schrödinger
    // equation `-iH * |psi>`, where `H` is the Hamiltonian.
    cudaq::super_op sup;
    // Apply `-iH * |psi>` superop
    sup +=
        cudaq::super_op::left_multiply(std::complex<double>(0.0, -1.0) * ham);
    auto result = cudaq::evolve(sup, dims, schedule, initialState, integrator,
                                {cudaq::spin_op::z(0)},
                                cudaq::IntermediateResultSave::All);
    // [End SuperOperator]
  }

  {
    auto H = spin_op::z(0) + spin_op::z(1);
    const cudaq::dimension_map dimensions = {{0, 2}, {1, 2}};
    cudaq::schedule schedule(cudaq::linspace(0.0, 1.0, 100), {"t"});
    cudaq::integrators::runge_kutta integrator(1, 0.001);
    // [Begin MPI]
    cudaq::mpi::initialize();
    // Initial state (expressed as an enum)
    auto psi0 = cudaq::InitialState::ZERO;

    // Run the simulation
    auto evolution_result =
        cudaq::evolve(H, dimensions, schedule, psi0, integrator);

    cudaq::mpi::finalize();
    // [End MPI]
  }
  return 0;
}

File: docs/sphinx/snippets/python/using/backends/dynamics.py
# ============================================================================ #
# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# Made up values - not sure what values are reasonable here.
#[Begin Transmon]
omega_z = 6.5
omega_x = 4.0
omega_d = 0.5

import numpy as np
from cudaq import spin, ScalarOperator

# Qubit Hamiltonian
hamiltonian = 0.5 * omega_z * spin.z(0)
# Add modulated driving term to the Hamiltonian
hamiltonian += omega_x * ScalarOperator(lambda t: np.cos(omega_d * t)) * spin.x(
    0)
#[End Transmon]

# Made up values - not sure what values are reasonable here.
t_final = 1.0
n_steps = 100

#[Begin Evolve]
import cudaq
import cupy as cp
from cudaq.dynamics import Schedule

# Set the target to our dynamics simulator
cudaq.set_target("dynamics")

# Dimensions of sub-systems: a single two-level system.
dimensions = {0: 2}

# Initial state of the system (ground state).
rho0 = cudaq.State.from_data(
    cp.array([[1.0, 0.0], [0.0, 0.0]], dtype=cp.complex128))

# Schedule of time steps.
steps = np.linspace(0, t_final, n_steps)
schedule = Schedule(steps, ["t"])

# Run the simulation.
evolution_result = cudaq.evolve(
    hamiltonian,
    dimensions,
    schedule,
    rho0,
    observables=[spin.x(0), spin.y(0), spin.z(0)],
    collapse_operators=[],
    store_intermediate_results=cudaq.IntermediateResultSave.ALL)
#[End Evolve]

#[Begin Plot]
get_result = lambda idx, res: [
    exp_vals[idx].expectation() for exp_vals in res.expectation_values()
]

import matplotlib.pyplot as plt

plt.plot(steps, get_result(0, evolution_result))
plt.plot(steps, get_result(1, evolution_result))
plt.plot(steps, get_result(2, evolution_result))
plt.ylabel("Expectation value")
plt.xlabel("Time")
plt.legend(("Sigma-X", "Sigma-Y", "Sigma-Z"))
#[End Plot]

# Made up values - not sure what values are reasonable here.
omega_c = 6 * np.pi
omega_a = 4 * np.pi
Omega = 0.5

#[Begin Jaynes-Cummings]
from cudaq import operators

hamiltonian = omega_c * operators.create(1) * operators.annihilate(1) \
                + (omega_a / 2) * spin.z(0) \
                + (Omega / 2) * (operators.annihilate(1) * spin.plus(0) + operators.create(1) * spin.minus(0))
#[End Jaynes-Cummings]

# Made up values - not sure what values are reasonable here.
omega = np.pi

#[Begin Hamiltonian]
# Define the static (drift) and control terms
H0 = spin.z(0)
H1 = spin.x(0)
H = H0 + ScalarOperator(lambda t: np.cos(omega * t)) * H1
#[End Hamiltonian]

#[Begin DefineOp]
import numpy
import scipy
from cudaq import operators, NumericType
from numpy.typing import NDArray


def displacement_matrix(
        dimension: int,
        displacement: NumericType) -> NDArray[numpy.complexfloating]:
    """
    Returns the displacement operator matrix.
    Args:
        displacement: Amplitude of the displacement operator.
            See also https://en.wikipedia.org/wiki/Displacement_operator.
    """
    displacement = complex(displacement)
    term1 = displacement * operators.create(0).to_matrix({0: dimension})
    term2 = numpy.conjugate(displacement) * operators.annihilate(0).to_matrix(
        {0: dimension})
    return scipy.linalg.expm(term1 - term2)


# The second argument here indicates the defined operator
# acts on a single degree of freedom, which can have any dimension.
# An argument [2], for example, would indicate that it can only
# act on a single degree of freedom with dimension two.
operators.define("displace", [0], displacement_matrix)


def displacement(degree: int) -> operators.MatrixOperatorElement:
    """
    Instantiates a displacement operator acting on the given degree of freedom.
    """
    return operators.instantiate("displace", [degree])


#[End DefineOp]

#[Begin Schedule1]
import cudaq

# Define a system consisting of a single degree of freedom (0) with dimension 3.
system_dimensions = {0: 3}
system_operator = displacement(0)

# Define the time dependency of the system operator as a schedule that linearly
# increases the displacement parameter from 0 to 1.
time_dependence = Schedule(numpy.linspace(0, 1, 100), ['displacement'])
initial_state = cudaq.State.from_data(
    numpy.ones(3, dtype=numpy.complex128) / numpy.sqrt(3))

# Simulate the evolution of the system under this time dependent operator.
cudaq.evolve(system_operator, system_dimensions, time_dependence, initial_state)
#[End Schedule1]

#[Begin Schedule2]
system_operator = displacement(0) + operators.squeeze(0)


# Define a schedule such that displacement amplitude increases linearly in time
# but the squeezing amplitude decreases, that is follows the inverse schedule.
def parameter_values(time_steps):

    def compute_value(param_name, step_idx):
        match param_name:
            case 'displacement':
                return time_steps[int(step_idx)]
            case 'squeezing':
                return time_steps[-int(step_idx + 1)]
            case _:
                raise ValueError(f"value for parameter {param_name} undefined")

    return Schedule(range(len(time_steps)), system_operator.parameters.keys(),
                    compute_value)


time_dependence = parameter_values(numpy.linspace(0, 1, 100))
cudaq.evolve(system_operator, system_dimensions, time_dependence, initial_state)
#[End Schedule2]

#[Begin SuperOperator]
import cudaq
from cudaq import spin, Schedule, RungeKuttaIntegrator
import numpy as np

hamiltonian = 2.0 * np.pi * 0.1 * spin.x(0)
steps = np.linspace(0, 1, 10)
schedule = Schedule(steps, ["t"])
dimensions = {0: 2}
# initial state
psi0 = cudaq.dynamics.InitialState.ZERO
# Create a super-operator that represents the evolution of the system
# under the Hamiltonian `-iH|psi>`, where `H` is the Hamiltonian.
se_super_op = cudaq.SuperOperator()
# Apply `-iH|psi>` super-operator
se_super_op += cudaq.SuperOperator.left_multiply(-1j * hamiltonian)
evolution_result = cudaq.evolve(se_super_op,
                                dimensions,
                                schedule,
                                psi0,
                                observables=[spin.z(0)],
                                store_intermediate_results=True,
                                integrator=RungeKuttaIntegrator())
#[End SuperOperator]

import cudaq
from cudaq import operators, spin, Schedule, RungeKuttaIntegrator

N = 4
dimensions = {}
for i in range(N):
    dimensions[i] = 2
g = 1.0
H = operators.zero()
for i in range(N):
    H += 2 * np.pi * spin.x(i)
    H += 2 * np.pi * spin.y(i)
for i in range(N - 1):
    H += 2 * np.pi * g * spin.x(i) * spin.x(i + 1)
    H += 2 * np.pi * g * spin.y(i) * spin.z(i + 1)

steps = np.linspace(0.0, 1, 200)
schedule = Schedule(steps, ["time"])

#[Begin MPI]
cudaq.mpi.initialize()

# Set the target to our dynamics simulator
cudaq.set_target("dynamics")

# Initial state (expressed as an enum)
psi0 = cudaq.dynamics.InitialState.ZERO

# Run the simulation
evolution_result = cudaq.evolve(
    H,
    dimensions,
    schedule,
    psi0,
    observables=[],
    collapse_operators=[],
    store_intermediate_results=cudaq.IntermediateResultSave.NONE,
    integrator=RungeKuttaIntegrator())

cudaq.mpi.finalize()
#[End MPI]
